### ISSUE TITLE
Potentially incorrect transformation of errors into HttpResponse

### ISSUE THREAD

**Issue by NavidJalali** [apache/pekko-grpc#392]
# TLDR: 
The way errors are turned into `HttpResponse` does not play well with AWS ALB.
# Long version:
We have several gRPC services behind an AWS ALB. While the Golang services seem to work just fine, all the Scala services seem to have an issue with returning errors.
For example when I call this behind the AWS ALB:
```bash
grpcurl -v -H "authorization: Bearer $ACCESS_TOKEN" -import-path ./core/src/main/protobuf -proto ./core/src/main/protobuf/banana.proto -d "{}" redacted.io:443 io.banana.v1beta1.BananaService/GetBanana
```
I get back:

> Resolved method descriptor:
> rpc GetBanana ( .io.banana.v1beta1.GetBananaRequest ) returns ( .io.banana.v1beta1.GetBananaResponse );
> 
> Request metadata to send:
> authorization: Bearer ???
> 
> Response headers received:
> (empty)
> 
> Response trailers received:
> content-length: 0
> content-type: application/grpc+proto
> date: Wed, 23 Oct 2024 14:13:22 GMT
> server: pekko-http/1.1.0
> x-ratelimit-limit: 0
> x-ratelimit-remaining: 0
> x-ratelimit-reset: 0
> Sent 1 request and received 0 responses
> ERROR:
>   Code: Unknown
>   Message: 

I get back an error message `Unknown` which is incorrect because if I port forward to this application, I can call it and see that it should in fact return a rich gRPC error:
```bash
grpcurl -v -import-path ./core/src/main/protobuf -proto ./core/src/main/protobuf/banana.proto --plaintext -d "{}" localhost:50051 io.banana.v1beta1.BananaService/GetBanana
```
Results in:
> Resolved method descriptor:
> rpc GetBanana ( .io.banana.v1beta1.GetBananaRequest ) returns ( .io.banana.v1beta1.GetBananaResponse );
> 
> Request metadata to send:
> (empty)
> 
> Response headers received:
> content-type: application/grpc+proto
> date: Wed, 23 Oct 2024 14:18:08 GMT
> server: pekko-http/1.1.0
> 
> Response trailers received:
> grpc-status-details-bin: CAUSGWNvdWxkIG5vdCBmaW5kIGJhbmFuYXMgOigaUQoodHlwZS5nb29nbGVhcGlzLmNvbS9nb29nbGUucnBjLkVycm9ySW5mbxIlCgp3ZSByYW4gb3V0Eglpby5iYW5hbmEaDAoGYmFuYW5hEgJubw==
> Sent 1 request and received 0 responses
> ERROR:
>   Code: NotFound
>   Message: could not find bananas :(
>   Details:
>   1)    {
>           "@error": "google.rpc.ErrorInfo is not recognized; see @value for raw binary message data",
>           "@type": "type.googleapis.com/google.rpc.ErrorInfo",
>           "@value": "Cgp3ZSByYW4gb3V0Eglpby5iYW5hbmEaDAoGYmFuYW5hEgJubw=="
>         }

This issue did not exist in the Golang services, so something about the way scala services were responding to the ALB was not standard.
The first breakthrough was finding out that [this issue was fixed in ZIO gRPC](https://github.com/scalapb/zio-grpc/issues/574) so by fixing to a snapshot that includes the fix, the errors were correctly propagated.
The issue links to [the gRPC specification section on retries](https://github.com/grpc/proposal/blob/master/A6-client-retries.md#when-retries-are-valid), particularly this paragraph:

> gRPC servers should delay the Response-Headers until the first response message or until the application code chooses to send headers. If the application code closes the stream with an error before sending headers or any response messages, gRPC servers should send the error in Trailers-Only.

Potentially, not doing this will confuse the AWS ALB in ways that result in the bad behavior mentioned above (Unknown error).

Testing out a scala application with the fix, I can see that indeed the header section is empty and the content type appears on the trailer.

> Request metadata to send:
> (empty)
> 
> Response headers received:
> (empty)
> 
> Response trailers received:
> content-type: application/grpc
> Sent 1 request and received 0 responses
> ERROR:
>   Code: NotFound
>   Message:

Looking even further with Wireshark I can see that the content-type is packed together with the grpc status.

<img width="1469" alt="image" src="https://github.com/user-attachments/assets/4019b4a3-79a5-4839-98b1-8bd9ef23a06d">

Inspecting the same for the Pekko gRPC service we can see that there are 2 separate sets of headers, the second one which includes the gRPC status and metadata is sent separately from the content-type etc.

<img width="1462" alt="image" src="https://github.com/user-attachments/assets/c61e0cae-0ec6-43eb-a5d1-0d070f12b790">

In a local version of Pekko gRPC I have changed the way failures are written to pack the trailers and the headers together. When checking on the AWS ALB, I can indeed see that it is correctly showing the errors now. 
```bash
grpcurl -v -H "authorization: Bearer $ACCESS_TOKEN" -import-path ./core/src/main/protobuf -proto ./core/src/main/protobuf/banana.proto -d "{}" redacted.io:443 io.banana.v1beta1.BananaService/GetBanana
```
Results in

> Response headers received:
> (empty)
> 
> Response trailers received:
> content-length: 0
> content-type: application/grpc+proto
> date: Wed, 23 Oct 2024 14:42:23 GMT
> grpc-status-details-bin: CAUSGWNvdWxkIG5vdCBmaW5kIGJhbmFuYXMgOigaUQoodHlwZS5nb29nbGVhcGlzLmNvbS9nb29nbGUucnBjLkVycm9ySW5mbxIlCgp3ZSByYW4gb3V0Eglpby5iYW5hbmEaDAoGYmFuYW5hEgJubw==
> server: awselb/2.0
> x-ratelimit-limit: 0
> x-ratelimit-remaining: 0
> x-ratelimit-reset: 0
> Sent 1 request and received 0 responses
> ERROR:
>   Code: NotFound
>   Message: could not find bananas :(
>   Details:
>   1)    {
>           "@error": "google.rpc.ErrorInfo is not recognized; see @value for raw binary message data",
>           "@type": "type.googleapis.com/google.rpc.ErrorInfo",
>           "@value": "Cgp3ZSByYW4gb3V0Eglpby5iYW5hbmEaDAoGYmFuYW5hEgJubw=="
>         }

I am not sure if the status quo is actually incorrect according to the specification, but this seems to be the same as libraries, and what AWS considers standard. I have limited understanding of the topic so I am happy to discuss further.

**Comment by NavidJalali** [comment]
Ran into some unrelated issue: https://github.com/apache/pekko-http/pull/622

**Comment by raboof** [comment]
I don't think sending errors in the trailing headers is wrong per the spec (https://github.com/grpc/proposal/blob/master/A6-client-retries.md#when-retries-are-valid seems like more of an optimization), but of course it's possible AWS ALB doesn't support this.

I think this should likely also be fixed on the AWS ALB side (there could also be situations where the error occurs after sending some data frames, in which case the error will have to be in the trailer). However, sending all headers in one headers section instead of splitting them between the regular and trailing headers where possible seems sensible.

**Comment by NavidJalali** [comment]
Thanks @raboof. I will try to contact AWS support and try to convey what you said faithfully, with my limited understanding of the underlying machinery 😄 
I will in the meantime try to set up a streaming test where there are a few frames sent before failing to see how that behaves, thank you for sharing your insights
